sm2/src/pke/encrypting.rs

```diff
@@ -0,0 +1,198 @@
+    /// then encodes the result in ASN.1 format.
+    pub fn encrypt_der(&self, msg: &[u8]) -> Result<Vec<u8>> {
+        self.encrypt_der_digest::<Sm3>(msg)
+    }
+
+    /// Encrypts a message using a specified digest algorithm.
+    pub fn encrypt_digest<D>(&self, msg: &[u8]) -> Result<Vec<u8>>
+    where
+        D: 'static + Digest + DynDigest + Send + Sync,
+    {
+        let mut digest = D::new();
+        encrypt(&self.public_key, self.mode, &mut digest, msg)
+    }
+
+    /// Encrypts a message using a specified digest algorithm and returns the result in ASN.1 format.
+    pub fn encrypt_der_digest<D>(&self, msg: &[u8]) -> Result<Vec<u8>>
+    where
+        D: 'static + Digest + DynDigest + Send + Sync,
+    {
+        let mut digest = D::new();
+        let cipher = encrypt(&self.public_key, self.mode, &mut digest, msg)?;
+        let digest_size = digest.output_size();
+        let (_, cipher) = cipher.split_at(1);
+        let (x, cipher) = cipher.split_at(32);
+        let (y, cipher) = cipher.split_at(32);
+        let (digest, cipher) = match self.mode {
+            Mode::C1C2C3 => {
+                let (cipher, digest) = cipher.split_at(cipher.len() - digest_size);
+                (digest, cipher)
+            }
+            Mode::C1C3C2 => cipher.split_at(digest_size),
+        };
+        Ok(Cipher {
+            x: Uint::from_be_slice(x),
+            y: Uint::from_be_slice(y),
+            digest,
+            cipher,
+        }
+        .to_der()
+        .map_err(elliptic_curve::pkcs8::Error::from)?)
+    }
+}
+
+impl From<PublicKey> for EncryptingKey {
+    fn from(value: PublicKey) -> Self {
+        Self::new(value)
+    }
+}
+
+/// Encrypts a message using the specified public key, mode, and digest algorithm.
+fn encrypt(
+    public_key: &PublicKey,
+    mode: Mode,
+    digest: &mut dyn DynDigest,
+    msg: &[u8],
+) -> Result<Vec<u8>> {
+    const N_BYTES: u32 = (Sm2::ORDER.bits() + 7) / 8;
+    let mut c1 = vec![0; (N_BYTES * 2 + 1) as usize];
+    let mut c2 = msg.to_owned();
+    let mut hpb: AffinePoint;
+    loop {
+        // A1: generate a random number 𝑘 ∈ [1, 𝑛 − 1] with the random number generator
+        let k = Scalar::from_uint(next_k(N_BYTES)).unwrap();
+
+        // A2: compute point 𝐶1 = [𝑘]𝐺 = (𝑥1, 𝑦1)
+        let kg = ProjectivePoint::mul_by_generator(&k).to_affine();
+
+        // A3: compute point 𝑆 = [ℎ]𝑃𝐵 of the elliptic curve
+        let pb_point = public_key.as_affine();
+        let s = *pb_point * Scalar::reduce(U256::from_u32(FieldElement::S));
+        if s.is_identity().into() {
+            return Err(Error);
+        }
+
+        // A4: compute point [𝑘]𝑃𝐵 = (𝑥2, 𝑦2)
+        hpb = (s * k).to_affine();
+
+        // A5: compute 𝑡 = 𝐾𝐷𝐹(𝑥2||𝑦2, 𝑘𝑙𝑒𝑛)
+        // A6: compute 𝐶2 = 𝑀 ⊕ t
+        kdf(digest, hpb, &mut c2)?;
+
+        // // If 𝑡 is an all-zero bit string, go to A1.
+        // if all of t are 0, xor(c2) == c2
```

Comment:
This comment seems a little odd

Reply:
I want to explain why we need to determine that xor(c2) == c2，but the description seems odd

Reply:
At the very least, there's a double `// //`

It seems like `If 𝑡 is an all-zero bit string` and `if all of t are 0` describe the same condition in slightly different ways.